### PR TITLE
mm: implement full cex api for simnet binance

### PR DESCRIPTION
**Pull request by buck54321**
In order to fully test the arb bots, we need full control over the markets as well as functional deposits and withdrawals. This branch updates the **testbinance** program to implement the full binance API. We can add coins and markets as needed.

Market are randomized in a small range around a basis price based on current coinpap fiat rates.

Did a fair amount of refactoring and fixed a couple of bugs.

Adds trace logging for arb-mm bot. Run `./walletpair.sh --simnet --no-embed-site --log MM=trace` to enable trace logging for just the mm package.

### CODE REVIEW

**Review by martonp — COMMENTED** [review]
Something is wrong with the orderbook syncing code. If you run `TestVWAP` in `binance_live_test.go` you can see in the logs that it's getting the snapshot over and over.

The new `testbinance` is clutch though.

File: client/cmd/testbinance/main.go [hunk 1]
```diff
@@ -7,955 +7,732 @@ package main
...
+			case <-ctx.Done():
+				return
+			}
+			runMarketTick()
+		}
+	}()
+
+	// Start a ticker to fill booked orders
+	go func() {
+		// 50% chance of filling all booked orders every 5 to 30 seconds.
```
[the remark this answers is not in the document]
Comment (buck54321) [reply client/cmd/testbinance/main.go:341]: I think there's a bunch of little tweaks we can make. Will add as needed.

**Review by martonp — APPROVED** [review]

**Review by JoeGruffins — APPROVED** [review]
A couple errors that popped up testing on simnet. From the binance server:
```
$ ./testbinance 
2024-04-26 15:38:41.438 [INF] C: Server listening on 0.0.0.0:37346
2024-04-26 15:38:41.438 [INF] C: Server listening on [::]:37346
2024-04-26 16:01:30.013 [ERR] TB: DELETE request received from user X-MBX-APIKEY for unknown order 164ea268145dafbfcb9500000001
2024-04-26 16:08:15.019 [ERR] TB: DELETE request received from user X-MBX-APIKEY for unknown order 164ea268145dafbfcb9500000002
2024-04-26 16:15:15.019 [ERR] TB: DELETE request received from user X-MBX-APIKEY for unknown order 164ea268145dafbfcb9500000003
2024-04-26 16:22:15.017 [ERR] TB: DELETE request received from user X-MBX-APIKEY for unknown order 164ea268145dafbfcb9500000004
2024-04-26 16:29:15.021 [ERR] TB: DELETE request received from user X-MBX-APIKEY for unknown order 164ea268145dafbfcb9500000005
2024-04-26 16:36:15.020 [ERR] TB: DELETE request received from user X-MBX-APIKEY for unknown order 164ea268145dafbfcb9500000006
```

And one for client:
```
2024-04-26 07:21:13.865 [ERR] MM[CEX-BinanceUS]: Error sending keep-alive request: http error (405) 405 Method Not Allowed
```

**Comment by buck54321** [comment]
> A couple errors that popped up testing on simnet. From the binance server:
> 
> ```
> $ ./testbinance 
> 2024-04-26 15:38:41.438 [INF] C: Server listening on 0.0.0.0:37346
> 2024-04-26 15:38:41.438 [INF] C: Server listening on [::]:37346
> 2024-04-26 16:01:30.013 [ERR] TB: DELETE request received from user X-MBX-APIKEY for unknown order 164ea268145dafbfcb9500000001
> 2024-04-26 16:08:15.019 [ERR] TB: DELETE request received from user X-MBX-APIKEY for unknown order 164ea268145dafbfcb9500000002
> 2024-04-26 16:15:15.019 [ERR] TB: DELETE request received from user X-MBX-APIKEY for unknown order 164ea268145dafbfcb9500000003
> 2024-04-26 16:22:15.017 [ERR] TB: DELETE request received from user X-MBX-APIKEY for unknown order 164ea268145dafbfcb9500000004
> 2024-04-26 16:29:15.021 [ERR] TB: DELETE request received from user X-MBX-APIKEY for unknown order 164ea268145dafbfcb9500000005
> 2024-04-26 16:36:15.020 [ERR] TB: DELETE request received from user X-MBX-APIKEY for unknown order 164ea268145dafbfcb9500000006
> ```
> 
> And one for client:
> 
> ```
> 2024-04-26 07:21:13.865 [ERR] MM[CEX-BinanceUS]: Error sending keep-alive request: http error (405) 405 Method Not Allowed
> ```

Were you just running an arb-mm bot for a while to get the unknown order errors? I have no idea how the users api key is being parsed as "X-MBX-APIKEY".

**Comment by JoeGruffins** [comment]
> Were you just running an arb-mm bot for a while to get the unknown order errors? I have no idea how the users api key is being parsed as "X-MBX-APIKEY".

Yeah... I put X-MBX-APIKEY as the api key I think... Maybe you don't need to put anything for simnet?

**Comment by buck54321** [comment]
I'll keep an eye on that missing order error. I didn't generate one overnight and it doesn't seem critical.

**Pull request merged**